x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/permission/DocumentPermissions.java

```diff
@@ -125,35 +136,25 @@ public BooleanQuery filter(
     ) throws IOException {
         if (hasDocumentLevelPermissions()) {
             evaluateQueries(SecurityQueryTemplateEvaluator.wrap(user, scriptService));
-            BooleanQuery.Builder filter;
-            if (evaluatedQueries != null && evaluatedLimitedByQueries != null) {
-                filter = new BooleanQuery.Builder();
-                BooleanQuery.Builder scopedFilter = new BooleanQuery.Builder();
-                buildRoleQuery(shardId, searchExecutionContextProvider, evaluatedLimitedByQueries, scopedFilter);
-                filter.add(scopedFilter.build(), FILTER);
+            assert listOfEvaluatedQueries != null : "evaluated queries must not be null";
+            assert false == listOfEvaluatedQueries.isEmpty() : "evaluated queries must not be empty";
 
-                buildRoleQuery(shardId, searchExecutionContextProvider, evaluatedQueries, filter);
-            } else if (evaluatedQueries != null) {
-                filter = new BooleanQuery.Builder();
-                buildRoleQuery(shardId, searchExecutionContextProvider, evaluatedQueries, filter);
-            } else if (evaluatedLimitedByQueries != null) {
-                filter = new BooleanQuery.Builder();
-                buildRoleQuery(shardId, searchExecutionContextProvider, evaluatedLimitedByQueries, filter);
-            } else {
-                assert false : "one of queries and limited-by queries must be non-null";
-                return null;
+            BooleanQuery.Builder filter = new BooleanQuery.Builder();
+            for (int i = listOfEvaluatedQueries.size() - 1; i > 0; i--) {
+                final BooleanQuery.Builder scopedFilter = new BooleanQuery.Builder();
+                buildRoleQuery(shardId, searchExecutionContextProvider, listOfEvaluatedQueries.get(i), scopedFilter);
+                filter.add(scopedFilter.build(), FILTER);
             }
+            // TODO: All role queries can be filters
```

Comment:
not sure i understand this TODO

Reply:
The existing logic is to have the limiting role (user) as filter. But the key's role as "should". So the query is something like:

```
{
  "bool": {
    "should": [ _key_queries ],
    "filter": [ { "bool": { "should": [ _Limiting_queries_ ] } } ]
  }
}
```

But the top level `should` clause is not necessary. Since intersection does not really care about order and does not treat any set of the queries specially, the TODO is proposing to change the above to:
```
{
  "bool": {
    "filter": [ 
      { "bool": { "should": [ _key_queries ] } } 
      { "bool": { "should": [ _Limiting_queries_ ] } } 
     ]
  }
}
```
This also helps simplifying the query building logic, i.e. not treating the 1st set any specially..